subprojects/file-watching/src/main/java/org/gradle/internal/watch/registry/impl/HierarchicalFileWatcherUpdater.java

```diff
@@ -58,96 +59,109 @@
-
-        determineAndUpdateWatchedHierarchies();
-    }
-
-    private void determineAndUpdateWatchedHierarchies() {
-        Set<Path> hierarchiesToWatch = determineHierarchiesToWatch();
-        updateWatchedHierarchies(hierarchiesToWatch);
+    public SnapshotHierarchy buildFinished(SnapshotHierarchy root) {
+        watchedHierarchiesFromPreviousBuild.clear();
+        watchedHierarchiesFromPreviousBuild.addAll(watchedHierarchies);
+        hierarchiesToWatchFromCurrentBuild.clear();
+
+        SnapshotHierarchy newRoot = recordHierarchiesToWatchAndRemoveUnwatchedSnapshots(
+            watchedHierarchies.stream().map(Path::toFile),
+            root,
+            (location, currentRoot) -> currentRoot.invalidate(location, SnapshotHierarchy.NodeDiffListener.NOOP)
+        );
+
+        determineAndUpdateWatchedHierarchies(newRoot);
+        LOGGER.warn("Watching {} directory hierarchies to track changes", watchedHierarchies.size());
+        LOGGER.info("Watched directory hierarchies: {}", watchedHierarchies);
+        return newRoot;
     }
 
-    private Set<Path> determineHierarchiesToWatch() {
-        Set<Path> directoriesToWatch = trackedDirectoriesForSnapshot.values().stream()
-            .map(trackedDirectory -> Stream.concat(knownRootProjectDirectoriesFromCurrentBuild.stream(), watchedRootProjectDirectoriesFromPreviousBuild.stream())
-                .filter(trackedDirectory::startsWith)
-                .findFirst()
-                .orElse(trackedDirectory))
+    private void determineAndUpdateWatchedHierarchies(SnapshotHierarchy root) {
+        Set<String> snapshotsAlreadyCoveredByOtherHierarchies = new HashSet<>();
+        Set<Path> hierarchiesWithSnapshots = Stream.concat(hierarchiesToWatchFromCurrentBuild.stream(), watchedHierarchiesFromPreviousBuild.stream())
+            .flatMap(hierarchyToWatch -> {
+                CheckIfNonEmptySnapshotVisitor checkIfNonEmptySnapshotVisitor = new CheckIfNonEmptySnapshotVisitor();
+                root.visitSnapshotRoots(hierarchyToWatch.toString(), new FilterAlreadyCoveredSnapshotsVisitor(checkIfNonEmptySnapshotVisitor, snapshotsAlreadyCoveredByOtherHierarchies));
+                if (checkIfNonEmptySnapshotVisitor.isEmpty()) {
+                    return Stream.empty();
+                }
+                return checkIfNonEmptySnapshotVisitor.containsOnlyMissingFiles()
+                    ? Stream.of(locationOrFirstExistingAncestor(hierarchyToWatch))
+                    : Stream.of(hierarchyToWatch);
+            })
             .collect(Collectors.toSet());
+        updateWatchedHierarchies(resolveHierarchiesToWatch(hierarchiesWithSnapshots));
+    }
 
-        return resolveHierarchiesToWatch(directoriesToWatch);
+    private Path locationOrFirstExistingAncestor(Path locationToWatch) {
+        if (Files.isDirectory(locationToWatch)) {
+            return locationToWatch;
+        }
```

Comment:
Could we use the VFS to determine this?

Reply:
No, because we don't snapshot the watchable directory. We only end up here if we don't have any information for the watchable directory in the VFS.